File: leetcode/1001-1999/1769-minimum-number-of-operations-to-move-all-balls-to-each-box/MinimumNumberOfOperationsToMoveAllBallsToEachBox.java
import java.util.Arrays;

public class MinimumNumberOfOperationsToMoveAllBallsToEachBox {
  public static void main(String[] args) {
    System.out.println("001011");
    System.out.println(Arrays.toString(new Solution().minOperations("001011")));
  }
}

class Solution {
  public int[] minOperations(String boxes) {
    int n = boxes.length();
    int[] operations = new int[n];

    for (int i = 1, balls = 0; i < boxes.length(); i++) {
      balls += boxes.charAt(i - 1) - '0';

      operations[i] += operations[i - 1] + balls;
    }

    System.out.println(Arrays.toString(operations));

    for (int i = n - 2, balls = 0, sum = 0; i >= 0; i--) {
      balls += boxes.charAt(i + 1) - '0';

      sum += balls;

      operations[i] += sum;
    }

    return operations;
  }
}

File: leetcode/1001-2000/1701-1800/1769-minimum-number-of-operations-to-move-all-balls-to-each-box/MinimumNumberOfOperationsToMoveAllBallsToEachBox.ts
function minOperations(boxes: string): number[] {
  const n = boxes.length;
  const balls = boxes.split("").map(c => c.charCodeAt(0) - '0'.charCodeAt(0));

  let operations: number[] = new Array(boxes.length).fill(0);

  for (let i = 1, accumulated = 0; i < n; i++) {
    accumulated += balls[i - 1];

    operations[i] = operations[i - 1] + accumulated;
  }

  for (let i = n - 2, accumulated = 0, sum = 0; i >= 0; i--) {
    accumulated += balls[i + 1];

    sum += accumulated;

    operations[i] = operations[i] + sum;
  }

  return operations;
};

console.log(minOperations("001011"));
